File: python solutions/min_depth_of_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None: return 0

        l = self.minDepth(root.left)
        r = self.minDepth(root.right)

        if root.left is None and root.right is None:
            return 1

        if root.left is None:
            return 1 + r
        
        if root.right is None:
            return 1 + l
        
        return min(l, r) + 1

File: ruby solutions/min_depth_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def min_depth(root)
    return 0 if root == nil
    l = min_depth(root.left)
    r = min_depth(root.right)
    return 1 if !root.left && !root.right
    return 1 + r if !root.left
    return 1 + l if !root.right
    return [l, r].min + 1
end